### PR TITLE
Refactor grading library to support projects

### PR DESCRIPTION
**Pull request by 0xDCA**
This PR refactors the grading library in order to support projects transparently. Only Python (2 and 3) is supported right now, but the rest of the languages will be adapted later.

### CODE REVIEW

**Comment by 0xDCA** [comment]
@amrondonp @milderhc

File: grading/base/grading/graders.py [hunk 1]
```diff
@@ -10,6 +11,7 @@
 import html
 import tempfile
 from . import projects
+from zipfile import ZipFile
 
 CODE_WORKING_DIR = 'student/'
```
Comment (0xDCA) [inline grading/base/grading/graders.py]: Oh, I actually left this duplicated in this file. The one that should be used is in projects.py

File: grading/base/grading/graders.py [hunk 2]
```diff
@@ -296,24 +298,36 @@ def handle_problem_action(problem_id, test_cases, language_name=None, options=No
...
-    project_factory = projects.get_factory_from_name(language_name)
-    project = project_factory.create_from_code(code)
+    project = None
+
+    if problem_type == 'code-multiple-languages':
+        project_factory = projects.get_factory_from_name(language_name)
+        project = project_factory.create_from_code(code)
+
+    elif problem_type == 'code-file-multiple-languages':
+        project_directory = os.path.join(CODE_WORKING_DIR, "code")
```
Comment (0xDCA) [inline grading/base/grading/graders.py]: I think it would be better if we had a subdirectory inside student. This prevents file name clashes in case the ZIP contains a file that is included in student/ as a resource for projects.

File: grading/base/grading/graders.py [hunk 3]
```diff
@@ -296,24 +298,36 @@ def handle_problem_action(problem_id, test_cases, language_name=None, options=No
...
     if language_name is None:
         language_name = input.get_input(problem_id + "/language")
+    problem_type = input.get_input(problem_id + "/type")
 
-    project_factory = projects.get_factory_from_name(language_name)
-    project = project_factory.create_from_code(code)
+    project = None
+
+    if problem_type == 'code-multiple-languages':
+        project_factory = projects.get_factory_from_name(language_name)
```
Comment (0xDCA) [inline grading/base/grading/graders.py]: Move project_factory out of this condition so it can be reused for both cases

File: grading/base/grading/projects.py [hunk 4]
```diff
@@ -99,7 +99,7 @@ class PythonProjectFactory(ProjectFactory):
 
     def __init__(self, main_file_name='main.py', python_binary='python', additional_flags=None):
         """
-        Initializes an instance of PythonRunner with the given options.
+        Initializes an instance of PythonProjectFactory with the given options.
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Thanks :)

File: grading/base/grading/projects.py [hunk 5]
```diff
@@ -128,14 +128,65 @@ def run(input_file):
         return LambdaProject(run_function=run)
 
 
+class JavaProjectFactory(ProjectFactory):
+    """
+    Implementation of ProjectFactory for Java.
+    """
+
+    def __init__(self, main_file_name='Main.java', source='1.7', sourcepath="src",
+                                                                 classpath="lib"):
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Correct the indentation so it's consistent with the rest of the file

File: grading/base/grading/projects.py [hunk 6]
```diff
@@ -128,14 +128,65 @@ def run(input_file):
...
+        self._source = source
+        self._sourcepath = sourcepath
+        self._classpath = classpath
+
+    def create_from_code(self, code):
+        def run(input_file):
+            javac_command = ["javac", self._main_file_name]
+            _run_in_sandbox(javac_command, stdin=input_file, cwd=CODE_WORKING_DIR)
+
+            java_command = ["java", self._main_file_name.replace(".java", "")]
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Don't do this. Instead, take a parameter in the constructor that tells you the name of the Main class (and create the file by appending '.java' to the end).

File: grading/base/grading/projects.py [hunk 7]
```diff
@@ -128,14 +128,65 @@ def run(input_file):
...
+
+        self._main_file_name = main_file_name
+        self._source = source
+        self._sourcepath = sourcepath
+        self._classpath = classpath
+
+    def create_from_code(self, code):
+        def run(input_file):
+            javac_command = ["javac", self._main_file_name]
+            _run_in_sandbox(javac_command, stdin=input_file, cwd=CODE_WORKING_DIR)
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Handle compilation errors and throw CompilationError. To do this, check the return code (non zero means there are errors) and use stderr as the compilation output

File: grading/base/grading/projects.py [hunk 8]
```diff
@@ -128,14 +128,65 @@ def run(input_file):
...
+
+            java_command = ["java", self._main_file_name.replace(".java", "")]
+            return _run_in_sandbox(java_command, stdin=input_file, cwd=CODE_WORKING_DIR)
+
+        with open(os.path.join(CODE_WORKING_DIR, self._main_file_name), 'w') as main_file:
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Please create a temporary directory for this to prevent name clashes (should be unlikely, but anyway)

File: grading/base/grading/projects.py [hunk 9]
```diff
@@ -128,14 +128,65 @@ def run(input_file):
...
+            main_file.write(code)
+
+        return LambdaProject(run_function=run)
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Can't we reuse the logic from create_from_directory here? It seems to me that it should be the same thing

File: grading/base/grading/projects.py [hunk 10]
```diff
@@ -128,14 +128,65 @@ def run(input_file):
...
+
+    def create_from_directory(self, directory):
+        build_directory = tempfile.mkdtemp(dir=directory)
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Why are you creating a temporary directory for builds? I think this should be always 'build' or something like that to make sure we use any caching the compiler may support.

File: grading/base/grading/projects.py [hunk 11]
```diff
@@ -128,14 +128,65 @@ def run(input_file):
...
+
+        def run(input_file):
+            build_directory = os.path.join(directory, "build")
+            if not os.path.exists(build_directory):
+                os.makedirs(build_directory)
+
+            javac_command = ["javac", "-source", self._source, "-d", "build", "-cp", self._classpath + "/*",
+                    "-sourcepath", self._sourcepath, os.path.join(self._sourcepath, self._main_file_name)]
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: I suggested compiling only the main file, but I think there is an edge case where this would not work. Some apps may use reflection to load some classes, rather than importing them (a lot of web frameworks do this). In this case, the compiler wouldn't compile all the files, and this will result in a runtime error when trying to load the classes. So I would suggest to compile all the java files in the source directory.

You can use the [glob](https://docs.python.org/3.5/library/glob.html) module with a pattern like "/path/to/source/**/*.java"

Comment (0xDCA) [inline grading/base/grading/projects.py]: Also, make sure you handle compilation errors too (or just reuse create_from_directory and handle it here)

File: grading/base/grading/projects.py [hunk 12]
```diff
@@ -128,14 +128,65 @@ def run(input_file):
...
+
+        return LambdaProject(run_function=run)
+
+
 _ALL_FACTORIES = {
     "python2": PythonProjectFactory(),
-    "python3": PythonProjectFactory(python_binary='python3')
+    "python3": PythonProjectFactory(python_binary='python3'),
+    "java7": JavaProjectFactory(),
+    "java8": JavaProjectFactory(source="1.8")
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: I think source 1.8 and target 1.8 are the defaults because the Java version installed in the container is Java 8. You need to do this for Java 7, but you need to pass the "-source" and "-target" flags to the compiler.

File: grading/base/grading/projects.py [hunk 13]
```diff
@@ -162,13 +162,11 @@ def run(input_file):
         return LambdaProject(run_function=run)
 
     def create_from_directory(self, directory):
-        build_directory = tempfile.mkdtemp(dir=directory)
+        build_directory = os.path.join(directory, "build")
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Oh, you actually did this. Ignore my previous comment about the temporary directory

File: grading/base/grading/projects.py (hunk 12 above)
Comment (milderhc) [inline grading/base/grading/projects.py]: Ok, I will swap the defaults. And I think there is no problem about not sending the "-target" flag, javac [documentation](http://docs.oracle.com/javase/8/docs/technotes/tools/windows/javac.html) says that the -target default value depends on the -source value, and one of the options says: "If the -source option is 1.7, then the value of the -target is option 1.8" so it will properly run in the 1.8 VM.

Comment (0xDCA) [inline grading/base/grading/projects.py]: Yes, you're right, I hadn't noticed that.

File: grading/base/grading/graders.py [hunk 14]
```diff
@@ -308,18 +306,19 @@ def handle_problem_action(problem_id, test_cases, language_name=None, options=No
...
 
     elif problem_type == 'code-file-multiple-languages':
-        project_directory = os.path.join(CODE_WORKING_DIR, "code")
+        unzip_directory = tempfile.mkdtemp(dir=projects.CODE_WORKING_DIR)
+        project_directory = os.path.join(unzip_directory, "code")
+
         with open(project_directory + ".zip", 'wb') as project_file:
             project_file.write(code)
-        ZipFile(project_directory + ".zip").extractall(path=CODE_WORKING_DIR)
+        ZipFile(project_directory + ".zip").extractall(path=unzip_directory)
```
Comment (0xDCA) [inline grading/base/grading/graders.py]: Please use ZipFile as a context manager (`with` statement)

File: grading/base/grading/projects.py [hunk 15]
```diff
@@ -143,36 +143,41 @@ def __init__(self, main_file_name='Main.java', source='1.7', sourcepath="src",
...
     def create_from_code(self, code):
-        def run(input_file):
-            javac_command = ["javac", "-source", self._source, self._main_file_name]
-            _run_in_sandbox(javac_command, stdin=input_file, cwd=CODE_WORKING_DIR)
-
-            java_command = ["java", self._main_file_name.replace(".java", "")]
-            return _run_in_sandbox(java_command, stdin=input_file, cwd=CODE_WORKING_DIR)
+        project_directory = tempfile.mkdtemp(dir=CODE_WORKING_DIR)
+        source_directory = os.path.join(project_directory, self._sourcepath)
+        if not os.path.exists(source_directory):
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: This is confusing. This directory should not exist before (you're using a a temporary directory). I think this case is unnecessary and misleading (if it does exist, which should not, it's better to fail fast: let it throw an exception and we'll know we have a problem)

File: grading/base/grading/projects.py [hunk 16]
```diff
@@ -143,36 +143,41 @@ def __init__(self, main_file_name='Main.java', source='1.7', sourcepath="src",
...
+        return self.create_from_directory(project_directory)
 
     def create_from_directory(self, directory):
         build_directory = os.path.join(directory, "build")
         if not os.path.exists(build_directory):
             os.makedirs(build_directory)
 
         def run(input_file):
+            task_absolute_path = os.getcwd()
+            os.chdir(directory)
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Why is this necessary?

File: grading/base/grading/projects.py [hunk 17]
```diff
@@ -143,36 +143,41 @@ def __init__(self, main_file_name='Main.java', source='1.7', sourcepath="src",
...
+            source_files = glob(os.path.join("**/*.java"), recursive=True)
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Why are you using os.path.join with a single path? I guess in this case it just returns the argument you passed.
Also, you're changing the working directory above, but why don't we use the absolute path in this glob? Something like `glob(os.path.join(directory, '**/*.java'), recursive=True)`

File: grading/base/grading/projects.py [hunk 18]
```diff
@@ -143,36 +143,41 @@ def __init__(self, main_file_name='Main.java', source='1.7', sourcepath="src",
...
+            os.chdir(task_absolute_path)
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: Why is this necessary? You are already using the CWD argument for _run_in_sandbox, so you shouldn't really need to change the working directory. I would prefer this method to be as stateless as possible, and this is a side effect.

File: grading/base/grading/projects.py [hunk 19]
```diff
@@ -143,36 +143,41 @@ def __init__(self, main_file_name='Main.java', source='1.7', sourcepath="src",
...
-                    "-sourcepath", self._sourcepath, os.path.join(self._sourcepath, self._main_file_name)]
-            _run_in_sandbox(javac_command, stdin=input_file, cwd=directory)
+                    "-sourcepath", self._sourcepath] + source_files
+            return_code, stdout, stderr = _run_in_sandbox(javac_command, stdin=input_file, cwd=directory)
+            if return_code != 0:
+                raise CompilationError(stderr)
 
-            java_command = ["java", "-cp" , "build" + ":" + self._classpath + "/*",
-                            self._main_file_name.replace(".java", "")]
+            java_command = ["java", "-cp" , "build" + ":" + self._classpath + "/*", self._main_class]
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: I'm confused. What is the colon (`:`) supposed to do? I couldn't find any reference in the documentation, and it does not seem to be a path.
Also, according to the documentation:

> A class path entry that contains * will not match class files. To match both classes and JAR files in a single directory foo, use either foo;foo/* or foo/*;foo. The order chosen determines whether the classes and resources in foo are loaded before JAR files in foo, or vice versa.

Do you think it's worth supporting class files too, or are we okay with just JAR support?

File: grading/base/grading/projects.py [hunk 20]
```diff
@@ -143,36 +143,41 @@ def __init__(self, main_file_name='Main.java', source='1.7', sourcepath="src",
             of the file where the code will be stored.
         """
 
-        self._main_file_name = main_file_name
+        self._main_class = main_class
+        self._main_file_name = self._main_class + ".java"
         self._source = source
```
Comment (0xDCA) [inline grading/base/grading/projects.py]: I find this name a bit confusing. Should we change it to source_version or something like that?

File: grading/base/grading/projects.py (hunk 17 above)
Comment (milderhc) [inline grading/base/grading/projects.py]: That will return the sources in the form "student/project_directory/code/src/File.java" and as we are compiling from the code directory we need the sources in the form "src/File.java", that's why I changed the working directory in that piece of code. And I don't know why I left "os.path.join("**/*.java")"... :v

File: grading/base/grading/projects.py [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+
+            javac_command = ["javac", "-source", self._source_version, "-d", "build", "-cp", self._classpath + "/*",
+                    "-sourcepath", self._sourcepath] + source_files
+            return_code, stdout, stderr = _run_in_sandbox(javac_command, stdin=input_file, cwd=directory)
+            if return_code != 0:
+                raise CompilationError(stderr)
+
+            classpath_entries = ["build", self._classpath, self._classpath + "/*"]
+            java_command = ["java", "-cp" , os.pathsep.join(classpath_entries), self._main_class]
+            return _run_in_sandbox(java_command, stdin=input_file, cwd=directory)
```
Comment (amrondonp) [inline grading/base/grading/projects.py]: @0xDCA Does this action creates another container? I am concern that if it that is not the case, the the compiled project wouldn't be present for excecution

Comment (0xDCA) [inline grading/base/grading/projects.py]: Yes, but the contents of the student/ folder are left unchanged

Comment (amrondonp) [inline grading/base/grading/projects.py]: Cool, I assume this is tested, in that case LGTM
